src/core/symbology-ng/qgsarrowsymbollayer.cpp

```diff
@@ -0,0 +1,589 @@
+
+  qreal headAngle = direction * atan( headSize / circleRadius );
+
+  QPainterPath path;
+
+  if ( headType == QgsArrowSymbolLayer::HeadDouble )
+  {
+    // the first head
+    path.moveTo( po );
+    path.lineTo( circlePoint( circleCenter, circleRadius + direction * headSize, angle_o + headAngle ) );
+
+    pathArcTo( path, circleCenter, circleRadius + direction * width / 2, angle_o + headAngle, angle_d - headAngle, direction );
+
+    // the second head
+    path.lineTo( circlePoint( circleCenter, circleRadius + direction * headSize, angle_d - headAngle ) );
+    path.lineTo( pd );
+    path.lineTo( circlePoint( circleCenter, circleRadius - direction * headSize, angle_d - headAngle ) );
+
+    pathArcTo( path, circleCenter, circleRadius - direction * width / 2, angle_d - headAngle, angle_o + headAngle, -direction );
+
+    // the end of the first head
+    path.lineTo( circlePoint( circleCenter, circleRadius - direction * headSize, angle_o + headAngle ) );
+    path.lineTo( po );
+  }
+  else if ( headType == QgsArrowSymbolLayer::HeadSingle )
+  {
+    path.moveTo( circlePoint( circleCenter, circleRadius + direction * startWidth / 2, angle_o ) );
+
+    spiralArcTo( path, circleCenter, angle_o, circleRadius + direction * startWidth / 2, angle_d - headAngle, circleRadius + direction * width / 2, direction );
+
+    // the arrow head
+    path.lineTo( circlePoint( circleCenter, circleRadius + direction * headSize, angle_d - headAngle ) );
+    path.lineTo( pd );
+    path.lineTo( circlePoint( circleCenter, circleRadius - direction * headSize, angle_d - headAngle ) );
+
+    spiralArcTo( path, circleCenter, angle_d - headAngle, circleRadius - direction * width / 2, angle_o, circleRadius - direction * startWidth / 2, -direction );
+
+    path.lineTo( circlePoint( circleCenter, circleRadius + direction * startWidth / 2, angle_o ) );
+  }
+  else if ( headType == QgsArrowSymbolLayer::HeadReversed )
+  {
+    path.moveTo( circlePoint( circleCenter, circleRadius + direction * width / 2, angle_o + headAngle ) );
+
+    spiralArcTo( path, circleCenter, angle_o + headAngle, circleRadius + direction * width / 2, angle_d, circleRadius + direction * startWidth / 2, direction );
+
+    path.lineTo( circlePoint( circleCenter, circleRadius - direction * startWidth / 2, angle_d ) );
+
+    spiralArcTo( path, circleCenter, angle_d, circleRadius - direction * startWidth / 2, angle_o + headAngle, circleRadius - direction * width / 2, - direction );
+
+    path.lineTo( circlePoint( circleCenter, circleRadius - direction * headSize, angle_o + headAngle ) );
+    path.lineTo( po );
+    path.lineTo( circlePoint( circleCenter, circleRadius + direction * headSize, angle_o + headAngle ) );
+    path.lineTo( circlePoint( circleCenter, circleRadius + direction * width / 2, angle_o + headAngle ) );
+  }
+
+  return path.toSubpathPolygons()[0];
+}
+
+void QgsArrowSymbolLayer::_resolveDataDefined( QgsSymbolV2RenderContext& context )
+{
+  bool ok;
```

Comment:
A good shortcut here would be to insert:

```
if ( !hasDataDefinedProperties() )
    return; // shortcut
```

This will speed up the render a bit by avoiding the multiple hasDataDefinedProperty checks (which will always return false)